tiledb/sm/filesystem/s3.cc

```diff
@@ -1332,37 +1345,66 @@ Status S3::write_multipart(
   const std::string uri_path(aws_uri.GetPath().c_str());
 
   // Take a lock protecting the shared multipart data structures
-  std::unique_lock<std::mutex> multipart_lck(multipart_upload_mtx_);
+  // Read lock to see if it exists
+  multipart_upload_mtx_.read_lock();
 
   auto state_iter = multipart_upload_states_.find(uri_path);
   if (state_iter == multipart_upload_states_.end()) {
-    // Delete file if it exists (overwrite) and initiate multipart request
-    bool exists;
-    RETURN_NOT_OK(is_object(uri, &exists));
-    if (exists) {
-      RETURN_NOT_OK(remove_object(uri));
-    }
+    // If the state is new, we must grab write lock, so unlock from read and
+    // grab write
+    multipart_upload_mtx_.read_unlock();
+    multipart_upload_mtx_.write_lock();
 
-    const Status st = initiate_multipart_request(aws_uri);
-    if (!st.ok()) {
-      return st;
-    }
+    // Since we switched locks we need to once again check to make sure another
+    // thread didn't create the state
+    auto state_iter = multipart_upload_states_.find(uri_path);
+    if (state_iter == multipart_upload_states_.end()) {
+      auto& path = aws_uri.GetPath();
+      std::string path_c_str = path.c_str();
+      MultiPartUploadState new_state;
+
+      assert(multipart_upload_states_.count(path_c_str) == 0);
+      multipart_upload_states_.emplace(
+          std::move(path_c_str), std::move(new_state));
+      MultiPartUploadState* state = &multipart_upload_states_.at(uri_path);
+      std::unique_lock<std::mutex> state_lck(state->mtx);
+      multipart_upload_mtx_.write_unlock();
+
+      // Delete file if it exists (overwrite) and initiate multipart request
+      bool exists;
+      RETURN_NOT_OK(is_object(uri, &exists));
+      if (exists) {
+        RETURN_NOT_OK(remove_object(uri));
+      }
 
-    state_iter = multipart_upload_states_.find(uri_path);
-    assert(state_iter != multipart_upload_states_.end());
+      const Status st = initiate_multipart_request(aws_uri, state);
+      if (!st.ok()) {
+        return st;
+      }
+      // Grab read lock, expected below outside the create
+      multipart_upload_mtx_.read_lock();
+    } else {
+      // If another thread switched state, switch back to a read lock
+      multipart_upload_mtx_.write_unlock();
+      multipart_upload_mtx_.read_lock();
+    }
   }
 
-  MultiPartUploadState* const state = &state_iter->second;
+  MultiPartUploadState* const state = &multipart_upload_states_.at(uri_path);
```

Comment:
we're no longer guaranteed that this still exists because we have released the write lock. the easiest thing to do here is to return an error if it no longer exists. the better thing to do would be to provide an interface to a rw-lock to synchronously downgrade to a read lock